sound/soc/intel/boards/sdw_rt711_rt1308_rt715.c

```diff
@@ -364,83 +366,263 @@ struct codec_dai dai_lists[] = {
+	const char *board;
+	int num_links;
+	int i, id = 0;
+	int val;
 
 #if IS_ENABLED(CONFIG_SND_SOC_HDAC_HDMI)
-	{
-		.name = "iDisp1",
-		.id = 5,
-		.init = hdmi_init,
-		.no_pcm = 1,
-		.dpcm_playback = 1,
-		SND_SOC_DAILINK_REG(idisp1_pin, idisp1_codec, platform),
-	},
-	{
-		.name = "iDisp2",
-		.id = 6,
-		.init = hdmi_init,
-		.no_pcm = 1,
-		.dpcm_playback = 1,
-		SND_SOC_DAILINK_REG(idisp2_pin, idisp2_codec, platform),
-	},
-	{
-		.name = "iDisp3",
-		.id = 7,
-		.init = hdmi_init,
-		.no_pcm = 1,
-		.dpcm_playback = 1,
-		SND_SOC_DAILINK_REG(idisp3_pin, idisp3_codec, platform),
-	},
+	board = dmi_get_system_info(DMI_BOARD_NAME);
+	if (strstr(board, "TigerLake"))
+		hdmi_num = 4;
+	else
+		hdmi_num = 3;
 #endif
-	{
-		.name = "SDW2-Playback",
-		.id = 3,
-		.init = second_spk_init,
-		.no_pcm = 1,
-		.dpcm_playback = 1,
-		.nonatomic = true,
-		SND_SOC_DAILINK_REG(sdw2_pin2, sdw2_codec, platform),
-	},
-};
+
+	if (snd_soc_acpi_find_machine(mach))
+		ssp_num = 1;
+
+	mach_params = &mach->mach_params;
+	sdw_num = get_sdw_dai_link_num(mach_params);
+
+	dev_dbg(dev, "sdw %d, ssp %d, hdmi %d", sdw_num, ssp_num, hdmi_num);
+
+	num_links = ssp_num + sdw_num + hdmi_num;
+	links = devm_kzalloc(dev, sizeof(struct snd_soc_dai_link) *
+			     num_links, GFP_KERNEL);
+	cpus = devm_kzalloc(dev, sizeof(struct snd_soc_dai_link_component) *
+			     num_links, GFP_KERNEL);
+	if (!links || !cpus)
+		goto devm_err;
+
+	/* SDW */
+	if (!sdw_num)
+		goto SSP;
+
+	link = mach_params->links;
+	while (link->num_adr) {
+		val = create_sdw_codec_dai(dev, links, id, cpus, link);
+		if (val < 0) {
+			dev_err(dev, "failed to create dai link %d", id);
+			goto devm_err;
+		}
+
+		id += val;
+		link++;
+	}
+
+SSP:
+	/* SSP */
+	if (ssp_num > 0) {
+		init_dai_link(links, id, "SSP2-Codec", cpus, "SSP2 Pin",
+			      rt1308_component, 1);
+
+		links[id].dpcm_playback = 1;
+		links[id].ops = &rt1308_i2s_ops,
+
+		id++;
+	}
+
+#if IS_ENABLED(CONFIG_SND_SOC_HDAC_HDMI)
+	/* HDMI */
+	if (hdmi_num > 0) {
+		val = sizeof(struct snd_soc_dai_link_component) * hdmi_num;
+		idisp_components = devm_kzalloc(dev, val, GFP_KERNEL);
+		if (!idisp_components)
+			goto devm_err;
+	}
+	for (i = 0; i < hdmi_num; i++) {
+		char *name, *cpu_name;
+
+		name = devm_kasprintf(dev, GFP_KERNEL,
+				      "iDisp%d", i + 1);
+		if (!name)
+			goto devm_err;
+
+		idisp_components[i].name = "ehdaudio0D2";
+		idisp_components[i].dai_name = devm_kasprintf(dev,
+							      GFP_KERNEL,
+							      "intel-hdmi-hifi%d",
+							      i + 1);
+		if (!idisp_components[i].dai_name)
+			goto devm_err;
+
+		cpu_name = devm_kasprintf(dev, GFP_KERNEL,
+					  "iDisp%d Pin", i + 1);
+		if (!cpu_name)
+			goto devm_err;
+
+		init_dai_link(links, id, name, cpus, cpu_name,
+			      idisp_components + i, 1);
+
+		links[id].init = hdmi_init;
+		links[id].dpcm_playback = 1;
+
+		id++;
+	}
+#endif
+
+	card->dai_link = links;
+	card->num_links = num_links;
+
+	return 0;
+devm_err:
+	return -ENOMEM;
```

Comment:
you could `return -Exxx;` in each error case and remove this label. Especially since your error code is fixed. So doing `return` everywhere directly you'd also be able to return any error code, even if you really only need ENOMEM.

Reply:
Thanks , I will refine it